tests/dht/test_providers.nim

```diff
@@ -28,23 +28,27 @@ import
 
 proc bootstrapNodes(
     nodecount: int,
-    bootnodes: openArray[SignedPeerRecord],
-    rng = keys.newRng()
-  ) : seq[(discv5_protocol.Protocol, keys.PrivateKey)] =
+    bootnodes: seq[SignedPeerRecord],
+    rng = keys.newRng(),
+    delay: int = 0
+  ) : Future[seq[(discv5_protocol.Protocol, keys.PrivateKey)]] {.async.} =
 
+  debug "---- STARTING BOOSTRAPS ---"
   for i in 0..<nodecount:
     let privKey = keys.PrivateKey.random(rng[])
     let node = initDiscoveryNode(rng, privKey, localAddress(20302 + i), bootnodes)
     node.start()
     result.add((node, privKey))
-  debug "---- STARTING BOOSTRAPS ---"
+    if delay > 0:
+      await sleepAsync(chronos.milliseconds(delay))
+
```

Comment:
In the tests, we don't ever seem to set `delay`, meaning it will always be `0`, so I'm curious how this altered the outcome of the tests. Maybe it was because `bootstrapNodes` was put in to the chronos event loop?

Reply:
it is not the delay that changed behaviour, but the async. The delay is still 0 as before.
The delay code is just there as I was experimenting with it and I had the code ready. We might use it in future tests.

Reply:
Fair enough, I guess I'm still not sure how making it async helped out when there is nothing being awaited inside `bootstrapNodes` (except `asyncSleep` when `delay > 0`, which never hits). My thinking was that simply marking it async allows it be processed as part of the chronos event loop. This could actually have something to do with `setupAll` assuming that the block is asynchronous...

Reply:
it's the `await` when calling `bootstrapNodes` which makes the difference. Nodes start and queue events (initial packets sent). The loop itself is not delayed, but when we return from the function, all these events start hitting the event loop.